Alternator/CDC: Implement streams support

elcallio opened the pull request:
Implementation of DynamoDB streams using Scylla CDC. 

Fixes #5065 

Initial, naive implementation insofar that it uses 1:1 mapping CDC stream to DynamoDB shard. I.e. there are a lot of shards. 

Includes tests verified against both local DynamoDB server and actual AWS remote one. 
Note: 
Because of how data put is implemented in alternator, currently we do not get "proper" `INSERT` labels for first write of data, because to CDC it looks like an update. The test compensates for this, but actual users might not like it.

psarna commented on alternator/rjson.hh:
```diff
@@ -80,6 +80,18 @@ using string_buffer = rapidjson::GenericStringBuffer<encoding>;
 using writer = rapidjson::Writer<string_buffer, encoding>;
 using type = rapidjson::Type;
 
+extern allocator the_allocator;
```
Is this pull request based on some other one? If so, what happened to that other one? I vaguely remember suggesting to wrap this allocator in `internal` namespace and you agreeing, but I don't see any trace of this discussion anywhere in here.

elcallio replied:
Last one was an RFC. But same series, comments dealt with an implementation complete and tested. But I apparently managed to drop that change when rebasing. I suck. Will fix.

nyh commented:
The first two commits are already in master, can you please rebase, or hit github over the head, or whatever it takes, to only have the relevant commits listed here? Thanks.

elcallio commented:
rebased.

elcallio commented on alternator/rjson.cc:
```diff
@@ -128,6 +128,18 @@ std::string print(const rjson::value& value) {
+rjson::missing_parameter::missing_parameter(std::string_view name) 
+    : error(format("JSON parameter {} not found", name))
+{}
+
```
The exceptions really only serve as message packing but with the extra benefit of being able to distinguish by type in caller catch. I fail to see _why_ you would not want this. The alternative is having copied code that formats the same message in mutiple places, alternatively a subroutine that is called instead of just using a different exception type. C-style. 
As for invalid_parameter, it says that the value could not be parsed as excepted by the (implicit) schema. While it is technically not a syntactical error (valid JSON), it is still a parse error (invalid JSON content).

elcallio replied:
Ok, I hear what you are saying here, and I admit, my using the exceptions above in actual alternator code is plain wrong. But the exception types are actually useful in them selves, mainly to unify message formatting (and keep it in slow path in cc-file) and potentially allow specialized catch (though that is not needed at this point, but...). 

I've changed this in next version so the rjson exceptions are purely rjson. Also added comments about how the existing use of rjson:error already breaks the separation in its message wording.

elcallio commented on alternator/rjson.hh:
```diff
@@ -163,6 +177,14 @@ void set_with_string_name(rjson::value& base, std::string_view name, rjson::stri
+std::enable_if_t<!std::is_constructible_v<string_ref_type, T>> 
+set(rjson::value& base, rjson::string_ref_type name, T&& member) {
+    rjson::value v;
+    v.Set(std::forward<T>(member), internal::the_allocator);
```
Not without going via ugly wormhole methods and std::function afaik.

elcallio replied:
fwiw, I made the allocator more hidden (function local extern), so should not affect anyone anymore.

elcallio commented on alternator/rjson.hh:
```diff
@@ -147,6 +151,16 @@ const rjson::value* find(const rjson::value& value, std::string_view name);
 
+template<typename T>
+T get(const rjson::value& value, std::string_view name) {
+    auto& v = get(value, name);
```
Sure, comments good. But how could you not be aware of the templating system in rjson? It was like the first thing I picked up on when reading the header...

elcallio commented on alternator/executor.cc:
```diff
@@ -157,31 +154,41 @@ static std::string lsi_name(const std::string& table_name, const std::string& in
-static std::string get_table_name(const rjson::value& request) {
-    const rjson::value& table_name_value = rjson::get(request, "TableName");
-    if (!table_name_value.IsString()) {
+std::optional<std::string> executor::get_table_name(const rjson::value& request, bool required) {
```
Sure. Same thing, different name.

elcallio commented on alternator/executor.hh:
```diff
@@ -78,6 +93,18 @@ public:
 
     static tracing::trace_state_ptr maybe_trace_query(client_state& client_state, sstring_view op, sstring_view query);
+
+    static sstring table_name(const schema&);
```
The neat thing about class-private functions it that they are class-private. I.e. if the intent is to use them for the purposes of the various calls in executor, we know no one will use them elsewhere. Of course this broke here, because of all the op-classes also using the free functions, and those not being class local. 

If you want me to move stuff to shared globals instead, then I ask for a header name (executor-internal.hh?).

elcallio replied:
Honestly, I am 100% ok with moving helpers into a module-private header + cc, but I suggest making this a later refactoring.

elcallio commented on alternator/streams.cc:
```diff
@@ -0,0 +1,186 @@
+#include "rmw_operation.hh"
+
+template<typename ValueType, typename T>
+struct from_string_helper {
```
rjson does. The rjsons system templates the rjson::value type, but in that context the line just says "to_string_view(rjson::value{})".

elcallio commented on alternator/streams.cc:
```diff
@@ -0,0 +1,186 @@
+            return boost::lexical_cast<T>(s.data(), s.size());
+        }
+    }
+    static ValueType& Set(ValueType&, T) = delete;
```
The type conversion template is expected to convert data for both data set with and without allocator. However, because we have a slightly more narrow use case, we limit it to the case _with_ allocator, i.e. rjson.hh:184. Otherwise we would have to produce values that are transient and do not need delete, which we probably cannot. 
Deleting the function will ensure that compiler will complain early if anyone tried to use the converter the wrong way.

elcallio commented on alternator/streams.cc:
```diff
@@ -0,0 +1,186 @@
+
+namespace alternator {
+
+// stream arn _has_ to be 37 or more characters long. ugh...
```
The spec literally says so. But sure, can add URL to this. (Btw, the actual boto3 driver already verifies this, so no ignoring the limit). 
Did you not write the UUID.hh code? One would think you were familiar with the string format of them. :-)

elcallio commented on alternator/streams.cc:
```diff
@@ -182,5 +182,277 @@ future<alternator::executor::request_return_type> alternator::executor::list_str
+    shard_id(const sstring&);
+
+    friend std::ostream& operator<<(std::ostream& os, const shard_id& id) {
+        return os << marker << id.table << std::hex
```
Because we only expect it to be used for lexical_cast? I am honestly on the fence on whether ostream operators should restore state, or never make assumptions on state to begin with. But maybe adding a stream state guard is prudent.

nyh replied:
What can I tell you, I never liked this operator<< stuff in C++ in the first place, and I'm not thrilled about bringing it into Alternator - but if it's stream.cc and you look after it (and take care not to print numbers in non-hex mode), it's fine with me :-)

elcallio replied:
I've added flag preservation at least.

elcallio commented on alternator/streams.cc:
```diff
@@ -182,5 +182,277 @@ future<alternator::executor::request_return_type> alternator::executor::list_str
+    static constexpr auto marker = 'H';
+
+    utils::UUID table;
+    db_clock::time_point time;
```
The generation time stamp. See CDC streams and their ID generation.

elcallio commented on alternator/streams.cc:
```diff
@@ -182,5 +182,277 @@ future<alternator::executor::request_return_type> alternator::executor::list_str
+        case stream_view_type::NEW_IMAGE: os << "NEW_IMAGE"; break;
+        case stream_view_type::OLD_IMAGE: os << "OLD_IMAGE"; break;
+        case stream_view_type::NEW_AND_OLD_IMAGES: os << "NEW_AND_OLD_IMAGES"; break;
+        default: throw std::invalid_argument(std::to_string(int(type)));
```
Easy: `auto type = stream_view_type(0xffffff)`. 
Would you prefer an assert? I find it bit of an overkill here, soft error seemed more fitting.

nyh replied:
The traditional exception to use in such cases is std::logic_error.

elcallio replied:
fair enough

elcallio commented on alternator/streams.cc:
```diff
@@ -182,5 +182,277 @@ future<alternator::executor::request_return_type> alternator::executor::list_str
+    if (opts.preimage() && opts.postimage()) {
+        type = stream_view_type::NEW_AND_OLD_IMAGES;
+    } else if (opts.preimage()) {
+        type = stream_view_type::OLD_IMAGE;
```
CDC itself only calculates postimage (from pre + delta) if postimage is enabled.

nyh replied:
Maybe CDC only calculates it, but isn't storing both on disk (in the CDC log) a waste of disk space? Wouldn't it be better to read the pre and delta and then *calculate*, on the fly, the post?
I don't mind that you add this as a "TODO" or "idea for optimization" for later, but I want to understand this...

elcallio replied:
While alternator has fairly simplistic column (i.e. no non-atomics), in general, the merge pre-image + delta is not 100% trivial. Note that we also allow post-image without pre-image, and so does dynamo streams. 
Calculating this on-the-fly has some benefits, but it would at the very least force us to make the alternator stream config persisted instead of implicit via CDC settings.

nyh replied:
Very well. Let's take it like this, but I will expect you to either remember this issue, or (better) create an issue in the bug tracker about the inefficiency in the current implementation saving all of preimage, delta and postimage - all on disk.

elcallio commented on alternator/streams.cc:
```diff
@@ -182,5 +182,277 @@ future<alternator::executor::request_return_type> alternator::executor::list_str
+    auto limit = rjson::get_opt<int>(request, "Limit").value_or(100); // according to spec
+    auto ret = rjson::empty_object();
+    auto stream_desc = rjson::empty_object();
+    // local java dynamodb server does _not_ validate the arn. But "real" one does. 
```
The validation is done in the parsing. I think it is worth noting where we are compatible with one and not the other (local/real), because one (me) running the local one and comparing with alternator might think "oh, we do it wrong". The comment will hopefully allow the code reader (me) to realize that it is in fact the local dynamo that behaves borked.

nyh replied:
Ok. My point is that we shouldn't run the local DynamoDB and nobody else should ever run the local DynamoDB, so how it behaves doesn't matter. But I guess a comment about the local DynamoDB doesn't hurt, though, even if we never plan to use it.

elcallio commented on alternator/streams.cc:
```diff
@@ -182,5 +182,277 @@ future<alternator::executor::request_return_type> alternator::executor::list_str
+    }
+ 
+    if (!schema || !bs || !is_alternator_keyspace(schema->ks_name())) {
+        std::throw_with_nested(api_error("ResourceNotFoundException", "Invalid StreamArn"));
```
Right. Remained from earlier control flow.

elcallio commented on alternator/streams.cc:
```diff
@@ -182,5 +182,277 @@ future<alternator::executor::request_return_type> alternator::executor::list_str
+    try {
+        shard_start = rjson::get_opt<shard_id>(request, "ExclusiveStartShardId");
+    } catch (...) {
+        // treated as nada. 
```
Maybe a bit short comment. If the start shard does not parse, it is simply silently not matched to anything, Setting limit=0 means we shortcut collecting anything and just return no results.

nyh replied:
Ok. As long as you tested (hopefully in one of the tests in Python) that this is how the real DynamoDB behaves - that giving it a crap ExclusiveStartShardId silently returns empty results but no error.

elcallio commented on alternator/streams.cc:
```diff
@@ -454,5 +454,144 @@ future<executor::request_return_type> executor::describe_stream(client_state& cl
+        case shard_iterator_type::AFTER_SEQUENCE_NUMBER: os << "AFTER_SEQUENCE_NUMBER"; break;
+        case shard_iterator_type::TRIM_HORIZON: os << "TRIM_HORIZON"; break;
+        case shard_iterator_type::LATEST: os << "LATEST"; break;
+        default: throw std::invalid_argument(std::to_string(int(type)));
```
Ok, but again, is assert the way to go here? Or do we have some less blunt tool to indicate borked values like this?

psarna replied:
`std::logic_error` comes to mind

elcallio replied:
But that is just another soft error. Or is @nyh:s comment really about the semantics of the exception name? If so, fair enough.

nyh replied:
Yes, I think std::logic_error is the traditional thing to do in this case.

elcallio commented on alternator/streams.cc:
```diff
@@ -454,5 +454,144 @@ future<executor::request_return_type> executor::describe_stream(client_state& cl
+        throw rjson::missing_parameter("SequenceNumber");
+    }
+    if (type >= shard_iterator_type::TRIM_HORIZON && seq_num) {
+        throw rjson::invalid_parameter("SequenceNumber", "Iterator of this type should not have sequence number");
```
Technically, invalid_pararmeter translates to a validationerror. It is just as specialization. Would you prefer it if the exception types simply changed namespace and parent? 
(One thing I really dislike is writing `throw api_error("ValidateionException", ...)` everywhere. Is should at the very least be a subtype `validation_exception`.

nyh replied:
I understand you dislike the throw api_error() everywhere, and yes, we could have done this as a subclass of api_error. BUT, it doesn't mean you need a subclass of rjson::error or that it needs to be in rjson.hh. It's not a JSON error at all, it is a DynamoDB API error. Yes, I know that rjson::error is translated by our server to an api_error, but we did it out of necessity (the actual rjson code might throw errors) and not because we want more of this rjson::error-throwing to happen.

Please note that for the DynamoDB API errors, you cannot just invent types - "ValidationException" is one of the few types of exceptions you can use (and which you should use in this case). You can pass different strings with this ValidationException, but you cannot really invent a new CalleException.

elcallio replied:
As I wrote above, I've ensured to separate exception types, and also created exception overloads for the dynamo errors.

elcallio commented on alternator/streams.cc:
```diff
@@ -593,4 +596,195 @@ future<executor::request_return_type> executor::get_shard_iterator(client_state&
+    }
+
+    if (!schema || !base || !is_alternator_keyspace(schema->ks_name())) {
+        std::throw_with_nested(api_error("ResourceNotFoundException", boost::lexical_cast<std::string>(iter.shard.table)));
```
same control flow... will fix.

elcallio commented on alternator/streams.cc:
```diff
@@ -593,4 +596,195 @@ future<executor::request_return_type> executor::get_shard_iterator(client_state&
+    static const bytes op_column_name = cdc::log_meta_column_name_bytes("operation");
+
+    auto key_names = boost::copy_range<std::unordered_set<std::string>>(
+        boost::range::join(std::move(base->partition_key_columns()), std::move(base->clustering_key_columns()))
```
You are not. But you are allowed to move the range objects returned. Somewhat pointless though.

elcallio commented on alternator/streams.cc:
```diff
@@ -786,5 +787,36 @@ future<executor::request_return_type> executor::get_records(client_state& client
+                break;
+            case stream_view_type::NEW_AND_OLD_IMAGES: 
+                opts.postimage(true);
+                opts.preimage(true);
```
Yes. But it is already done for us by CDC is postimage is one. And I do not want to duplicate that logic as well.

nyh replied:
Is this "I don't want to duplicate this logic yet" (which is perfectly fine, but please add a FIXME comment), or "we should never do it"? If it's the latter, aren't we bothered that the implication is that we store as much as double the information on-disk?

elcallio replied:
Well, this is a CDC design choice. (See comment above). Technically, I think alternator data storage is so simple that doing on-the-fly post image here would be feasible, but then if we change the way alternator stores data, we would run into issues. And of course, having post-image without pre-image more or less creates the same situation. 
I was instead thinking that we should make it possible to have CDC pre- and/or postimage _without_ delta, to minimize unneeded data for alternator. 
But that is a later optimization.

nyh replied:
Yes, it's a later optimization, but we'll need to remember it somewhere (as an issue, FIXME, or something).

elcallio commented on alternator/executor.cc:
```diff
@@ -1854,11 +1857,42 @@ std::unordered_set<std::string> calculate_attrs_to_get(const rjson::value& req,
+    const std::vector<bytes_opt>& result_row,
+    const std::unordered_set<std::string>& attrs_to_get,
+    rjson::value& item,
+    bool include_all_embedded_attributes) 
```
Because I don't need/want the full monty of the wrapping function, and `include_all_embedded_attributes` is because we literally want everything in `executor::ATTRS_COLUMN_NAME`. Without having to figure the key names out. See comment in GetRecords on why we don't want `attrs_to_get` empty.

elcallio commented on alternator/executor.cc:
```diff
@@ -986,6 +984,54 @@ future<executor::request_return_type> executor::create_table(client_state& clien
     });
 }
 
+future<executor::request_return_type> executor::update_table(client_state& client_state, tracing::trace_state_ptr trace_state, service_permit permit, rjson::value request) {
```
Sure.

elcallio commented on docs/alternator/alternator.md:
```diff
@@ -134,8 +134,13 @@ implemented, with the following limitations:
-  (change data capture), but the "DynamoDB Streams" API is not yet supported.
+* Implemented via [CDC](https://docs.scylladb.com/using-scylla/cdc/)
+  (change data capture). The Alternator server responed to all DynamoDB Streams 
+  API calls, but is limited to a single stream per table.
```
Dynamo does according to docs allow more than one stream. It did not quite work smoothly for me when testing, but I might have done it wrong.

nyh replied:
I still don't understand what you mean. DynamoDB allows you to "enable a stream" on a table. And that's it! You can't enable a second stream... 
Maybe you mean that it is possible to enable a stream, then disable it, and then re-enable it with different paramters and (I guess) get a different stream ID. All of the above needs UpdateTable which we don't support yet. Is that what your code doesn't allow? If so, please explain this more clearly.

elcallio replied:
Changed wording

elcallio commented on docs/alternator/alternator.md:
```diff
@@ -134,8 +134,13 @@ implemented, with the following limitations:
+  API calls, but is limited to a single stream per table.
+  Note that because of how Scylla CDC operates, there is a time window
+  between data being written to a table and it being visible via
+  GetRecords calls (30s).
```
We should not document the internals here. My take is measure, and probably move to a unified "(CDC) stream grouping scheme" that we use across Alternator, Java driver, etc etc. But need solidifying that idea. Until then, it is hidden by abstraction here. 
We do actually sort of implement parent shards, though it is imprecise.

nyh replied:
My intention wasn't to document the internals, but to mention things that might surprise the user when they are different from DynamoDB. So if a user sees 5 (I don't know... do they?) shards on a small DynamoDB test but 500,000 shards with Alternator, we should warn them. 

I didn't realise about the children shards - I didn't see this in the code and I assumed it was missing... So you're saying the code as you sent now works with topology changes and the client will be able to read also data from the new news? We should definitely have a dtest in the future that tests this.

kbr- commented on cdc/generation.hh:
```diff
@@ -113,6 +114,22 @@ public:
     const std::vector<token_range_description>& entries() const;
 };
 
+class streams_version {
```
Could you document this class a bit?
Specifically, what does the "version" refer to?

elcallio replied:
I can. It is a version of the set of streams.

kbr- replied:
How are these sets of streams versioned?
Who/what defines the versioning?

If we have `streams_version sv1, sv2`, when is the version of `sv2` greater than `sv1`?

Also, maybe `versioned_streams` would be a better name?

elcallio replied:
You version them. Version == generation. I.e. a topology change creates a new version of the stream id set. 
Maybe `stream_topology_generation` is a better name. Feel free to suggest.

elcallio replied:
I avoided the `topology` name though, to not confuse it with something actually used in the topology calculation. This is purely for query of CDC_DESC.

kbr- replied:
Ok, so it's versioned by the generation's timestamp.

This is basically a CDC generation but with the information about mapping of streams to token ranges forgotten. I.e. what we give to the client in CDC_DESC for querying streams.

kbr- replied:
Maybe `generation_stream_set`?

`streams_version` is fine, but a comment would be nice in any case.

elcallio commented on alternator/executor.cc:
```diff
@@ -1854,11 +1857,42 @@ std::unordered_set<std::string> calculate_attrs_to_get(const rjson::value& req,
         const cql3::selection::selection& selection,
         const query::result& query_result,
-        std::unordered_set<std::string>&& attrs_to_get) {
+        const std::unordered_set<std::string>& attrs_to_get) {
```
It really bugged me, but it technically also prevented retaining an `attrs_to_get` set through a lambda (where value members become const), and then use without copying.

elcallio commented on test/alternator/test_streams.py:
```diff
@@ -0,0 +1,373 @@
+                print(streams)
+                if not streams.get('Streams'):
+                    return
+                print("Stream(s) lingering. Sleep 30s...")
```
Note that these delays will only happen if running against actual dynamo. There, creating and deleting streams is quite expensive, and requires waiting. For alternator and local dynamo we will never do the delays.

nyh replied:
Ok. I don't mind if the tests aren't perfect yet, we can improve them later. However, I don't want tests taking more than a fraction of a second when running against Alternator. If there are such tests, I think we should "skip" them (with the pytest tag) in the meanwhile and then think what to do about it.
By the way, even if DynamoDB is slow, it doesn't mean we need to wait a multiple of 30 seconds - whic means a 31 second wait gets rounded to 60 seconds. We could sleep in multiples of, say, 2 seconds. But I don't insist. I really care more about the performance against the local Alternator - which we run and rerun every time - and not the performance against DynamoDB (which we only use to develop the test).

elcallio commented on test/alternator/test_streams.py:
```diff
@@ -0,0 +1,373 @@
+            table.delete()
+            return
+        except ClientError as ce:
+            if ce.response['Error']['Code'] == 'ResourceInUseException':
```
It happens if you have streams enabled. That is why.

nyh replied:
I don't understand... The DeleteTable documentation doesn't say anything about ResourceInUseException being returned when the stream is... being... what?

elcallio replied:
If the stream is enabled but not yet available you can get exceptions when deleting the table (not ready).

elcallio commented on test/alternator/test_streams.py:
```diff
@@ -0,0 +1,373 @@
+                if arn != None:
+                    return arn;
+        print("Stream not available. Sleep 5s...")
+        time.sleep(5)
```
You will not get 5 seconds, because we need to create new tables most of the time. But again, the loop-delay above only happens when running against real dynamo. Which responds very slowly to stream state changes.

nyh replied:
Yes, creating tables is a bummer, but even that takes on my laptop about 0.1 seconds (maybe tables with CDC are slower, but probably not significantly slower). What I don't want to see is that table-creation only takes 0.1 seconds but the test already started to sleep for 5 seconds.

elcallio commented on test/alternator/test_streams.py:
```diff
@@ -0,0 +1,373 @@
+# args. This will try to check if we are in fact running against 
+# this test server, and if so, just raise the error here and be done 
+# with it. All this just so we can run through the tests on 
+# aws, scylla and local. 
```
I found it useful to be able to use the local one. It does some compatibility checking, and more importantly allows one to figure out driver quirks before actually using real dynamo.

nyh replied:
Ok. I still think using the real Dynamo is easy enough as it is that the local one isn't necessary and worse - a distraction - because it caused you to try to support (in your code and tests) two DynamoDB implementations instead of one. But I don't strongly disagree that you leave it, since you already did the work.

elcallio commented on test/alternator/test_streams.py:
```diff
@@ -0,0 +1,373 @@
+def test_create_streams_wrong_type(dynamodb, dynamodbstreams, test_table):
+    with pytest.raises(ClientError, match='ValidationException'):
+        test_table.update(StreamSpecification={'StreamEnabled': True, 'StreamViewType': 'Fisk'});
+        test_table.update(StreamSpecification={'StreamEnabled': False});
```
Yes. Just kept the disable in case the test fails.

elcallio commented on test/alternator/test_streams.py:
```diff
@@ -0,0 +1,373 @@
+        with pytest.raises(ClientError, match='ValidationException'):
+            streams = dynamodbstreams.list_streams(TableName=table.name, ExclusiveStartStreamArn='kossaapaaasdafsdaasdasdasdasasdasfadfadfasdasdas')
+            assert 'Streams' in streams
+            assert not streams['Streams'] # empty
```
Leftovers from copy paste.

elcallio replied:
No, sorry, it is supporting the local java server case. Adding comment.

elcallio commented on alternator/streams.cc:
```diff
@@ -593,4 +596,195 @@ future<executor::request_return_type> executor::get_shard_iterator(client_state&
+    auto iter = rjson::get<shard_iterator>(request, "ShardIterator");
+    auto limit = rjson::get_opt<size_t>(request, "Limit").value_or(1000);
+
+    if (limit == 0) {
```
Unfortunately this cannot be easily tested, because the driver already checks this. But I've changed it to an error condition in scylla.

elcallio commented on test/alternator/test_streams.py:
```diff
@@ -0,0 +1,373 @@
+        assert desc['StreamDescription']['Shards'][0]['SequenceNumberRange'].get('StartingSequenceNumber')
+
+def test_describe_nonexistent_stream(dynamodb, dynamodbstreams):
+    with pytest.raises(ClientError, match='ResourceNotFoundException|ValidationException'):
```
Again, local. But changed to verify explicitly.

elcallio commented on test/alternator/test_streams.py:
```diff
@@ -0,0 +1,373 @@
+                            assert oldimage.get('a1')
+                            assert oldimage.get('a2')
+                            assert oldimage['a1']['S'] == val
+                            assert oldimage['a2']['S'] == val2
```
Do secondary indices affect the base table layout in alternator? Are those not separate tables?
I agree that is would be good to force a base table that has both attrs map and real columns (except the keys) and verify this, but I am unsure on how to create this. 
I would maybe want to postpone adding this to the test (added comment).

nyh replied:
In the current implementation, when secondary indexes are not involved, just the two key columns (partition and sort) are real Scylla columns - all the rest are stuffed into one ":attrs" column (a map). However, with LSI or GSI, we have additional indexed column(s), and this column(s) are real Scylla columns as well (because of limitations of our materialized views implementation, which doesn't know to take values from the map and needs to index a real column). So if the streams code expects that all of the columns except the key are always in ":attrs", it won't work when there are LSI or GSI.
Sure, you can add a TODO and check this later. Even better (but I don't insist) is to write a test, and if it fails on Alternator, leave it as "xfail" - these are much better than a TODO.

elcallio commented:
Pushed new version that should (hopefully) address all comments above. In some cases I have just added TODO:s and/or comments on why things are what they are. 
* Fixed exception types and usage (rjson exception only in rjson, alternator exceptions in alternator). 
* Fixed exception usage (logic error, alternator exceptions etc). 
* Fixed some checks and conditions. 
* Cleaned up names and interfaces of exported executor functions. Removed some as well.  
* Lots of comments added based on review
* Added some test cases + comments + removed spurious prints. Full set of tests now take ~18s in release when running scylla. Not fast, but not shop stopper either. 
* Updated title and descriptions of several commits

slivne commented:
@haaawk / @nyh / @psarna / @kbr- - please review we want to merge this

nyh commented:
> @haaawk / @nyh / @psarna / @kbr- - please review we want to merge this

I will review it now. I also want to merge it, and improve on it later, as such a huge PR is a big pain for me to review, and I'm sure an even bigger pain for Calle.

kbr- commented on cdc/log.hh:
```diff
@@ -129,7 +129,11 @@ enum class operation : int8_t {
 bool is_log_for_some_table(const sstring& ks_name, const std::string_view& table_name);
-seastar::sstring log_name(const seastar::sstring& table_name);
+schema_ptr get_base_table(const database&, const sstring& ks_name, const std::string_view& table_name);
+schema_ptr get_base_table(const database&, const schema&);
```
wrong commit

elcallio replied:
drat. fixed.

kbr- commented on cdc/log.cc:
```diff
@@ -352,13 +352,24 @@ static bool is_log_name(const std::string_view& table_name) {
+    return get_base_table(db, s.ks_name(), s.cf_name());
+}
+
+schema_ptr get_base_table(const database& db, const sstring& ks_name, const std::string_view& table_name) {
```
why `const string_view&` and not just `string_view`?

elcallio replied:
Just copy-paste from `is_log_for_some_table`. Fixed it to be consistent.

kbr- commented on cdc/generation.hh:
```diff
@@ -113,6 +114,31 @@ public:
 };
 
+/**
+ * The set of a streams for a single topology version/generation
```
```suggestion
 * The set of streams for a single topology version/generation
```

elcallio replied:
Fixed

elcallio commented on alternator/executor.cc:
```diff
@@ -933,8 +933,6 @@ future<executor::request_return_type> executor::create_table(client_state& clien
         return make_ready_future<request_return_type>(api_error("ValidationException", "SSESpecification: configuring encryption-at-rest is not yet supported."));
     }
-    // We don't yet support streams (CDC), but a StreamSpecification asking
-    // *not* to use streams should be accepted:
```
Fixed

nyh commented on alternator/executor.cc:
```diff
@@ -1861,11 +1864,57 @@ std::unordered_set<std::string> calculate_attrs_to_get(const rjson::value& req,
+ * We could change the behaviour to simply include all values 
+ * from this column if the `ATTRS_COLUMN_NAME` is explicit in 
+ * `attrs_to_get`, but I am scared to do that now in case 
+ * there is some corner case in existing code.
```
This comment is based on a misunderstanding on what attrs_to_get_is... It is a list of DynamoDB attributes. The "ATTRS_COLUMN_NAME" is just an implmentation detail in Alternator. 
I think this function would have been better off without this comment at all which casts doubts more than it explains. The name "include_all_embedded_attribute" is quite meaningless, there is no notion of "embedded attributes" - the same attribute may be in the map (ATTRS_COLUMN_NAME) or as a separate column if it is a key of the base table or LSI f the table - it's not clear why we need a separate option to return the first type of attributes but not the second - and the name "embedded" doesn't quire explain the distinction.

elcallio replied:
As explained in the code: The problem with the attributes in the :attr column is that they are not visible in the scylla schema. I want to get all data without having to figure out what data is called in the map (the ones in columns we know already, implicitly). I am open to modifying the API here, but the basic requirement remains, and I felt this was least resistance.

nyh replied:
If you want to get *all* columns you could have called the flag that - no need to list separately the columns in the schema and another flag for those not in the schema (which you called "embedded"). But heck, I think we already return all the columns if the list of required attributes is empty, didn't we???
I'll fix in a followup patch, if I can (maybe as I try to fix it I'll find I'm misunderstanding something?)

elcallio replied:
But I don't explicitly want _all_ columns. I want those mentioned and those embedded in :attr. The problem with empty attr_map is that it then don't return any _actual_ column values. So if we have mixed real column/embedded values we would have to call twice. And merge.

elcallio commented on alternator/error.hh:
```diff
@@ -44,7 +54,24 @@ public:
+};
+}
+
+using validation_exception = detail::t_api_error<detail::exception_type::ValidationException>;
```
A factory function is the exact same thing, but C-style (and with a lot of boiler plate duplication for the various errors) and without the benefit of separate type (i.e. discernable on catch/log level). I fail to see how this is better. Seriously, the template here is literally the simplest and least error prone you can get: Cannot use the wrong type + all aspects handled by code completion/intellisense.

nyh replied:
I'll just have to send a followup patch with my suggestion. I think it is exactly the opposite of what you said - there will be less boiler-plate code than you had to add, not more, code will be clearer, simpler, and usages will be clearer - when you write api_error::something its clearer that it is a subclass of api_error than just something_exception which can be any C++ exception thing. 
 don't think that "C-style" is a good way to judge anything. There is no "opposite of C-style" - there is no C++ style. C++ style was once upon a time heavily based on OOP, today it's heavily generics (templates), and you actually used both techniques in a place that doesn't need either :-) When I'll send a followup patch you can see what I meant.

elcallio replied:
Ok, you mean `static [noreturn] void api_error::throw_validation_exception(sstring msg)`. Fair enough, but that has the downside of a) not being able to _catch_ specific error. b) if the throw is in a factory function which is not inline thw compiler can miss optimization ops. 
Can of course use  `static api_error api_error::throw_validation_exception(sstring msg)`, but still lacks a)
Personally, I prefer specific types, since that is how exception handling was meant to be used.

nyh replied:
I will send a follow-up patch which undoes your templated hierarchy and returns a **single** type of exception, `api_error`, with multiple factory functions to build them (instead of explicit strings).
The idea of having a single type of `api_error` exception wasn't because I like C so much (?). It was because there is only one way to return an "api error" in the HTTP response - containing an error "name" and JSON content - and this is what "api_error" contains. Nobody is expected to "catch" the different types of api_errors - they are designed to be directly sent to the user, as is. This is why producers of these errors throw them. Code which throws exceptions which should be caught - should throw other types of exceptions - not api_errors (if such an exception goes all the way back to the user, we translate it to an "internal error" api_error).

Yes, what I want is something like `static api_error api_error::validation(sstring msg)`. Like this - no need for the extra long name "make_validation_exception" - the name `api_error::validation` is clear and shorter. 

Again, I'll do this as a followup patch and you can see what it looks like, and object to it (if you still do) then.

elcallio commented on db/config.cc:
```diff
@@ -739,6 +739,7 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , alternator_address(this, "alternator_address", value_status::Used, "0.0.0.0", "Alternator API listening address")
     , alternator_enforce_authorization(this, "alternator_enforce_authorization", value_status::Used, false, "Enforce checking the authorization header for every request in Alternator")
     , alternator_write_isolation(this, "alternator_write_isolation", value_status::Used, "", "Default write isolation policy for Alternator")
+    , alternator_streams_time_window_s(this, "alternator_streams_time_window_s", value_status::Used, 10, "CDC query confidence window for alternator streams")
```
Suggestions welcome

nyh replied:
Can come later. Let's keep this name for now.

elcallio commented:
See #6838

nyh commented:
> See #6838

Thanks!

nyh commented:
> See #6838

Thanks!

nyh commented:
@elcallio unfortunately I can't merge this, it doesn't compile:
```
alternator/streams.cc: In lambda function:
alternator/streams.cc:723:138: error: ‘class utils::time_estimated_histogram’ has no member named ‘_count’; did you mean ‘count’?
```
Can you please rebase? I think it is some histogram changes that @amnonh committed recently.

elcallio commented:
rebased and fixed the error.

nyh commented:
@elcallio unfortunately this version compiles, but doesn't pass the Alternator test (`test/alternator/tun`, or equivalently `test.py --mode=dev alternator/run`) - two tests fail with unclear problems, and I suspect this is just because test/alternator/run does not enable experimental mode so CDC is not enabled. I'll send a patch to change change `test/alternator/run`, which we will need to commit first - before I can commit this - otherwise tests will start failing :-(

However, @elcallio, I also need an answer from you: Do I need any other option to make the test work (and efficient)? Maybe making the read delay lower?

Later we will also need to add clear errors when Streams API is used and experimental mode is disabled. I don't want these weird silent failures. I'll do that as a followup patch, we can live with this problem in the meantime.

nyh commented:
@elcallio please add the following patch to your series. I can send it separately outside your series, but it will look very strange.
```patch
Author: Nadav Har'El <nyh@scylladb.com>
Date:   Wed Jul 15 20:20:36 2020 +0300

    alternator test: enable experimental CDC
    
    In the script test/alternator/run, which runs Scylla for the Alternator
    tests, add the "--experimental-features=cdc" option, to allow us testing
    the streams API whose implementation requires the experimenal CDC feature.
    
    Signed-off-by: Nadav Har'El <nyh@scylladb.com>

diff --git a/test/alternator/run b/test/alternator/run
index d294d39f1..dffdfbac0 100755
--- a/test/alternator/run
+++ b/test/alternator/run
@@ -85,6 +85,7 @@ ln -s "$SCYLLA" "$SCYLLA_LINK"
         --alternator-enforce-authorization=1 \
         --alternator-write-isolation=always_use_lwt \
         --developer-mode=1 \
+        --experimental-features=cdc \
         --ring-delay-ms 0 --collectd 0 \
         --smp 2 -m 1G \
         --overprovisioned --unsafe-bypass-fsync 1 \
```

nyh commented:
@elcallio unfortunately, even with this patch adding experimental CDC, one test still fails. Please reproduce the following failure by running "test/alternator/run test_streams.py". This "run" script runs Scylla for you (takes less than a second to boot!) and runs your pytest - it's easier than running Scylla yourself and then "pytest" against it.
Anyway, here is the failure I get:
```
=================================== FAILURES ===================================
_______________________________ test_get_records _______________________________

dynamodb = dynamodb.ServiceResource()
dynamodbstreams = <botocore.client.DynamoDBStreams object at 0x7fa91e72de80>

    def test_get_records(dynamodb, dynamodbstreams):
        # TODO: add tests for storage/transactionable variations and global/local index
        with create_stream_test_table(dynamodb, StreamViewType='NEW_AND_OLD_IMAGES') as table:
            arn = wait_for_active_stream(dynamodbstreams, table)
    
            p = 'piglet'
            c = 'ninja'
            val = 'lucifers'
            val2 = 'flowers'
>           table.put_item(Item={'p': p, 'c': c, 'a1': val, 'a2': val2})
test_streams.py:316: 
...
E           botocore.exceptions.ClientError: An error occurred (Internal Server Error) when calling the PutItem operation (reached max retries: 3): Internal server error: std::runtime_error (cdc::metadata::get_stream: could not find any CDC stream (current time: 2020/07/15 17:26:36). Are we in the middle of a cluster upgrade?)
```

Can you please look into it? Note that this is a single-node cluster, which obviously is not in the middle of a cluster upgrade.

Also, I get one unrelated warning from Python, if you can and you're sending a new version anyway, please fix it. Thanks.

nyh commented:
@elcallio sorry for the spam, but I really want to commit this series.
Would you object that I modify your patches in-place to simply mark the one currently-failing test as xfail, and you will look into it later and send a followup patch to fix it?

nyh commented:
I'm marking the failing test with "xfail" and committing.

nyh commented:
Committed 61f52da9b1c53da6dad25cc90f22a98df56e752c (merging your code), c4497bf770eb7b9d8a523b929f7b7f373109797d (add experimental cdc to test/alternator/run) and dcf9c888a2c2a892b0fd3d99e061fcceaec58b31 (xfail one test). Let's continue discussing futher improvements and fixes on the mailing list.

elcallio commented:
> 
> @elcallio unfortunately this version compiles, but doesn't pass the Alternator test (`test/alternator/tun`, or equivalently `test.py --mode=dev alternator/run`) - two tests fail with unclear problems, and I suspect this is just because test/alternator/run does not enable experimental mode so CDC is not enabled. I'll send a patch to change change `test/alternator/run`, which we will need to commit first - before I can commit this - otherwise tests will start failing :-(
>
Ugh, sorry. Did not realize CDC was off in unit test scripts. I ran things manually (faster). 
 
> However, @elcallio, I also need an answer from you: Do I need any other option to make the test work (and efficient)? Maybe making the read delay lower?
> 
It is already reduced to 10s. The whole streams suite should take 30s or less in release mode on Scylla (way longer against AWS, but...). 
I think that is a fair compromise. 

> Later we will also need to add clear errors when Streams API is used and experimental mode is disabled. I don't want these weird silent failures. I'll do that as a followup patch, we can live with this problem in the meantime.

Very true. Did not think about that at all actually (living in a CDC bubble).

elcallio commented:
> 
> 
> @elcallio unfortunately, even with this patch adding experimental CDC, one test still fails. Please reproduce the following failure by running "test/alternator/run test_streams.py". This "run" script runs Scylla for you (takes less than a second to boot!) and runs your pytest - it's easier than running Scylla yourself and then "pytest" against it.
> Anyway, here is the failure I get:
> 
> ```
> =================================== FAILURES ===================================
> _______________________________ test_get_records _______________________________
> 
> dynamodb = dynamodb.ServiceResource()
> dynamodbstreams = <botocore.client.DynamoDBStreams object at 0x7fa91e72de80>
> 
>     def test_get_records(dynamodb, dynamodbstreams):
>         # TODO: add tests for storage/transactionable variations and global/local index
>         with create_stream_test_table(dynamodb, StreamViewType='NEW_AND_OLD_IMAGES') as table:
>             arn = wait_for_active_stream(dynamodbstreams, table)
>     
>             p = 'piglet'
>             c = 'ninja'
>             val = 'lucifers'
>             val2 = 'flowers'
> >           table.put_item(Item={'p': p, 'c': c, 'a1': val, 'a2': val2})
> test_streams.py:316: 
> ...
> E           botocore.exceptions.ClientError: An error occurred (Internal Server Error) when calling the PutItem operation (reached max retries: 3): Internal server error: std::runtime_error (cdc::metadata::get_stream: could not find any CDC stream (current time: 2020/07/15 17:26:36). Are we in the middle of a cluster upgrade?)
> ```
> 
> Can you please look into it? Note that this is a single-node cluster, which obviously is not in the middle of a cluster upgrade.
> 
> Also, I get one unrelated warning from Python, if you can and you're sending a new version anyway, please fix it. Thanks.

Sorry for not being online until you already commited. :-)
The above worked fine for me at the time of last rebase. Will investigate why it suddenly fails.

nyh commented:
> Ugh, sorry. Did not realize CDC was off in unit test scripts. I ran things manually (faster).

Just a comment about running Alternator's tests - using the script `test/alternator/run` (the same script which `test.py` ends up running) is actually the fastest way to run the Alternator tests against Alternator: It boots Scylla with all sorts of "clever" parameters in less than one second, and then runs the tests you want against it. You only need to compile Scylla before, not run it. Basically, you can do:
```
ninja build/dev/scylla
test/alternator/run test_streams.py | less
```
to run just the streams test, or just "test/alternator/run" to run all the 500+ tests (this takes about 30 seconds).

Yes, sometimes it is indeed convenient for different reasons to run Scylla yourself, and then run pytest against the already running Scylla - but faster it usually isn't.

> > However, @elcallio, I also need an answer from you: Do I need any other option to make the test work (and efficient)? Maybe making the read delay lower?
> 
> It is already reduced to 10s. The whole streams suite should take 30s or less in release mode on Scylla (way longer against AWS, but...).
> I think that is a fair compromise.

I agree. It actually takes just 3 seconds on my laptop so this is perfectly acceptable! Thanks.

elcallio commented:
> 
> 
> > Ugh, sorry. Did not realize CDC was off in unit test scripts. I ran things manually (faster).
> 
> Just a comment about running Alternator's tests - using the script `test/alternator/run` (the same script which `test.py` ends up running) is actually the fastest way to run the Alternator tests against Alternator: It boots Scylla with all sorts of "clever" parameters in less than one second, and then runs the tests you want against it. You only need to compile Scylla before, not run it. Basically, you can do:

Yeah, I know. But I use those same parameters on the command line and keep the instance running, then I can run individual tests against the instance without waiting at all. Also helps when running debug.
